### PR TITLE
View Trips Page Real time Updates

### LINKED ISSUE #62: Implement Real Time Updates for all Trip Collaborators
### Issue summary

We want real time listening to Firestore so that a user's trips are always up to date.

### Expected behavior

The addition of real time listening with [onShapshot](https://firebase.google.com/docs/firestore/query-data/listen) should allow fetching the most up to date trips where a user is a collaborator whenever a new update to the Trips collection is made.

### Actual behavior

Currently when a new trip is created/edited, only the creator of the trip can automatically see the trip on their 'View Trips' page (using the [key prop](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key) on TripsContainer). For other users, they must refresh the page to see any new trips where they are collaborators.

### Steps to reproduce (including any preconditions)

{Fill in here}

### Screenshot/screencast of this problem (if possible)

{Fill in here or remove (leave header)}

### Your environment

Include version numbers

- OS: {Fill in here}
- Browser: {Fill in here}

### PR DESCRIPTION
**Pull request by zghera**
### What is a quick description of the change?

Enable real time updates such that any collaborator's update to a trip (add, edit, or delete) is immediately reflected in the local view of the trip for all collaborators. 

### Is this fixing an issue?

fixes #62 

### Are there more details that are relevant?

The main addition of this PR is to the `ViewTrips` directory with the following files:
- `trips-container.js`:  
  - `get` was replaced with `onSnapshot` when querying trips in order to use the real-time update features Firestore has to offer. 
  - Additionally, `queryuserTrips` and `serveTrips` were combined and moved into `componentDidMount`.This was done due to the constraints set by `onSnapshot`: both the querySnapshot handling callback and the error callback must be passed into `onSnapshot` rather than `onSnapshot` returning a promise.
- `index.js`, `save-trip-modal.js`, `trip.js`, and `delete-trip-button.js`: Remove all functions, variables, props, and references related `refreshTripContainer` (including TODOs to fix #62).

### Check lists (check `x` in `[ ]` of list items)

- [  ] Test written/updating
- [  ] Tests passing
- [x] Coding style (indentation, etc)

As of now, integration (end-to-end) tests were deemed non-critical. Thus, functions related to react and firestore will be tested at a later time. Unit tests were not written for this PR as there was no (original) "logical" functions written to complete this feature.

### Any additional comments?

GIF showing real-time updates for a shared trip between users step2020.53.test1@gmail.com and step2020.53.test2@gmail.com:
real-time

Note: I could not use two chrome windows because only one user can be signed into SLURP at a time in all instances of chrome. So the left window is Firefox ESR (that is why it looks kinda gross).

### CODE REVIEW

File: frontend/src/components/ViewTrips/trips-container.js [hunk 1]
```diff
@@ -86,19 +43,39 @@ class TripsContainer extends React.Component {
...
+                  key={doc.id}
+                />
+              )
+          );
+
+          this.setState({ trips: tripsContainer });
+        }, (error) => {
+          let errorElement = getErrorElement(error);
+          this.setState({ trips: errorElement });
+        });
```
Comment (ananyeet) [inline frontend/src/components/ViewTrips/trips-container.js:78]: Is this all that's required for the realtime stuff? whoa.

Comment (zghera) [reply frontend/src/components/ViewTrips/trips-container.js:78]: Yeah I was surprised at how little difference there is.

File: frontend/src/components/ViewTrips/trips-container.js [hunk 2]
```diff
@@ -63,7 +18,13 @@ function serveTrips(querySnapshot, handleEditTrip, refreshTripsContainer) {
 function getErrorElement(error) {
   return new Promise(function(resolve) {
     console.log(`Error in Trips Container: ${error}`);
-  resolve(( <div><p>Error: Unable to load your trips.</p></div> ));
+    resolve(
+      <div>
+        <p>Oops, it looks like we were unable to load your trips.
+                      Please wait a few minutes and try again.
+        </p>
+      </div>
```
Comment (keiffer01) [inline frontend/src/components/ViewTrips/trips-container.js:26]: One of my PRs created an `Errors` directory, so once that's merged do you think you could turn this function into a component and place it there? An issue could be created in the meantime.

**Comment by zghera** [comment]
> Wow, the real-time feature is really nice for such a relatively small change to the code, pretty nice! FYI Chrome incognito windows might help when testing with two different accounts simultaneously.

Yeah it is super nice! I tried to use incognito windows but it did not work when trying to log in. I got this error:
"This browser is not supported or 3rd party cookies and data may be disabled"

File: frontend/src/components/ViewTrips/trips-container.js (hunk 2 above)
Comment (zghera) [reply frontend/src/components/ViewTrips/trips-container.js:26]: Sure thing. I will add an issue right now.

File: frontend/src/components/Utils/time.test.js [hunk 3]
```diff
@@ -120,16 +120,15 @@ test('other time 24h format', () => {
...
+test('ISODate empty input tests', () => {
   const testDate = new Date(Date.UTC(2020, 7, 23, 2, 3, 2, 4)).getTime();
   const expected = '2020-08-23';
   expect(utils.getISODate(testDate, '')).toBe(expected);
   expect(utils.getISODate(testDate, null)).toBe(expected);
-})
- })
-})
+});
+
```
Comment (zghera) [inline frontend/src/components/Utils/time.test.js:131]: FYI if you see this change @Hiram-Silvey , the tests were broken so I made this change in this branch. But I am going to open a separate PR so it is not randomly in this branch.

**Pull request merged**